### PR TITLE
Setup Vetomint

### PR DESCRIPTION
**Pull request by junha1**

### CODE REVIEW

File: consensus/src/vetomint.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+///
+/// Note that there is no cryptography-related info here, because it's
+/// the lower layer's responsibility to verifiy and refine the raw messages (containing such cryptography-related info) into this abstracted data.
+/// Also all the identifiers (for blocks and validators) become integer indices here, and
+/// the lower layer will keep the mapping from the actual data to the indices.
+#[derive(Debug, Clone, PartialEq)]
+pub enum ConsensusEvent {
+    BlockProposal {
+        proposal: BlockIdentifier,
+        proposer: ValidatorIndex,
```
Comment (byeongjee) [inline consensus/src/vetomint.rs:22]: Do we need proposer field? I think it's already known

File: consensus/src/vetomint.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    },
+    NilPrevote {
+        signer: ValidatorIndex,
+        round: usize,
+    },
+    NilPrecommit {
+        signer: ValidatorIndex,
+        round: usize,
+    },
+    Timer {
```
Comment (byeongjee) [inline consensus/src/vetomint.rs:45]: Is this used to notify the consensus engine that the current timeout is expired?
Then why do we need time field?

File: consensus/src/vetomint.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        time: Timestamp,
+    },
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum ConsensusResponse {
+    /// Creation of the actual proposal is not the role of the consensus; the lower layer will take care of it.
+    CreateAndBroadcastProposal {
```
Comment (byeongjee) [inline consensus/src/vetomint.rs:53]: What event triggers this reponse?
perhaps (the last) precommit?

File: consensus/src/vetomint.rs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        round: usize,
+    },
+    BroadcastPrevote {
```
Comment (byeongjee) [inline consensus/src/vetomint.rs:56]: I'm not sure about the abstract layer of `ConsensusResponse`, but if you want to include `BroadcastPrevote`, you should include `StopBroadcastingPrevote` as well.

File: consensus/src/vetomint.rs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+pub enum ConsensusEvent {
+    BlockProposal {
+        proposal: BlockIdentifier,
+        proposer: ValidatorIndex,
+        round: usize,
+        /// Whether the operator of this consensus is in favor of this proposal.
+        favor: bool,
+    },
+    Prevote {
+        proposal: BlockIdentifier,
```
Comment (byeongjee) [inline consensus/src/vetomint.rs:28]: Given a fixed round, proposal is fixed as well. Do we need this field?

File: consensus/src/vetomint.rs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+pub type Timestamp = i64;
+
+/// An event that (potentially) triggers a state transition of `StateMachine`.
+///
+/// Note that there is no cryptography-related info here, because it's
+/// the lower layer's responsibility to verifiy and refine the raw messages (containing such cryptography-related info) into this abstracted data.
+/// Also all the identifiers (for blocks and validators) become integer indices here, and
+/// the lower layer will keep the mapping from the actual data to the indices.
+#[derive(Debug, Clone, PartialEq)]
+pub enum ConsensusEvent {
```
Comment (byeongjee) [inline consensus/src/vetomint.rs:19]: Will there be some external message filter/queue responsible for handling messages from different heights?

Comment (junha1) [reply consensus/src/vetomint.rs:19]: Yes

File: consensus/src/vetomint.rs (hunk 1 above)
Comment (junha1) [reply consensus/src/vetomint.rs:22]: Double proposals are possible

File: consensus/src/vetomint.rs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    BlockProposal {
+        proposal: BlockIdentifier,
+        proposer: ValidatorIndex,
+        round: usize,
+        /// Whether the operator of this consensus is in favor of this proposal.
```
[the remark this answers is not in the document]
Comment (junha1) [reply consensus/src/vetomint.rs:24]: I mean 'this node'. Will fix.

File: consensus/src/vetomint.rs (hunk 5 above)
Comment (junha1) [reply consensus/src/vetomint.rs:28]: The proposal's not fixed, as the earlier comment answered.

File: consensus/src/vetomint.rs (hunk 2 above)
Comment (junha1) [reply consensus/src/vetomint.rs:45]: I'll add timestamp for all the other events

File: consensus/src/vetomint.rs (hunk 3 above)
Comment (junha1) [reply consensus/src/vetomint.rs:53]: yes

File: consensus/src/vetomint.rs (hunk 4 above)
Comment (junha1) [reply consensus/src/vetomint.rs:56]: All information during a single height should be endlessly broadcasted

**Review by byeongjee — APPROVED** [review]

**Review by yeomjh00 — APPROVED** [review]
LGTM

**Review by P1n3tree — APPROVED** [review]

**Pull request merged**